service/consul/datacenterWatch.go

```diff
@@ -1,109 +1,199 @@
 package consul
 
 import (
+	"context"
 	"errors"
+	"sync"
 	"time"
 
 	"github.com/go-kit/kit/log"
 	"github.com/go-kit/kit/log/level"
+	"github.com/mitchellh/mapstructure"
+	"github.com/xmidt-org/argus/chrysom"
+	"github.com/xmidt-org/argus/model"
 	"github.com/xmidt-org/webpa-common/logging"
 	"github.com/xmidt-org/webpa-common/service"
 )
 
-//datacenterWatcher checks if datacenters have been updated, based on an interval
-//If they have, then the datacenterWatcher will update the Instancers dictionary
+//DatacenterWatcher checks if datacenters have been updated, based on an interval
 type datacenterWatcher struct {
-	watchInterval time.Duration
-	logger        log.Logger
-	shutdown      chan struct{}
-	environment   Environment
-	options       Options
+	logger              log.Logger
+	environment         Environment
+	options             Options
+	inactiveDatacenters map[string]bool
+	chrysomClient       *chrysom.Client
+	consulWatchInterval time.Duration
+	lock                sync.RWMutex
+}
+
+type datacenterFilter struct {
+	Name     string `mapstructure:"name"`
+	Inactive bool   `mapstructure:"inactive"`
 }
 
 var (
 	defaultLogger = log.NewNopLogger()
 )
 
 func newDatacenterWatcher(logger log.Logger, environment Environment, options Options) (*datacenterWatcher, error) {
-	if options.DatacenterWatchInterval == 0 {
-		return nil, errors.New("interval cannot be 0")
-	}
 
 	if logger == nil {
 		logger = defaultLogger
 	}
 
-	return &datacenterWatcher{
-		watchInterval: options.DatacenterWatchInterval,
-		logger:        logger,
-		shutdown:      make(chan struct{}),
-		environment:   environment,
-		options:       options,
-	}, nil
+	if options.DatacenterWatchInterval <= 0 {
+		//default consul interval is 5m
+		options.DatacenterWatchInterval = time.Duration(5 * time.Minute)
+	}
 
-}
+	datacenterWatcher := &datacenterWatcher{
+		consulWatchInterval: options.DatacenterWatchInterval,
+		logger:              logger,
+		options:             options,
+		environment:         environment,
+		inactiveDatacenters: make(map[string]bool),
+	}
 
-func (i *datacenterWatcher) start() {
-	go i.watchDatacenters()
-}
+	if options.ChrysomConfig != nil && options.ChrysomConfig.PullInterval > 0 {
 
-func (i *datacenterWatcher) stop() {
-	close(i.shutdown)
-}
+		if environment.Provider() == nil {
```

Comment:
We only use this `Provider()` when we use chrysom?

Reply:
Read further and the answer is yes...I think we should move this, but only because I think we should add a metric to this struct (see another comment below).  Could add a comment for now, explaining that only the chrysom client has metrics.